[Sol] Step3-02-2 속성 변경 기능 추가 및 Plane 구조체 유닛테스트 작성

Hansolkkim opened the pull request:
## 실행 화면
![SS 2022-03-07 PM 06 18 45](https://user-images.githubusercontent.com/92504186/157002715-f7827f7a-063d-4448-acf4-a16ce61eb832.gif)
## 작업 목록
- [x] Refactoring
  - [x] 튜플형 매개변수에 Label 명시
  - [x] Alpha, BackgroundColor 생성시 값 제한하도록 failable 생성자 이용
- [x] 사각형 뷰와 Rectangle 객체의 매칭을 위해 사각형 뷰에 해당하는 RectangleView 선언
- [x] 사각형 뷰 터치를 구현하기 위해 GestureRecognizer 사용
- [x] 우측의 배경색 버튼을 누르면 랜덤으로 Rectangle 객체 및 매칭되는 사각형 뷰의 BackgroundColor 랜덤하게 변경
- [x] 우측의 UISlider를 이용해 Rectangle 객체 및 매칭되는 사각형 뷰의 Alpha 변경
- [x] 우측의 (-),(+) 버튼을 이용해 단계별로 Rectangle 객체의 Alpha 변경
- [x] Rectangle이 선택되면 그에 맞는 배경색 버튼, UISlider를 설정하기 위한 Delegate 패턴 적용
- [x] 유닛테스트 작성 
## 고민과 해결
- 이전에는 타입에서 값을 제한하기 보다는 제한 값을 넘는 값이 오면 최소,최댓값으로 고정시켜주도록 했습니다. 그런데 저번 리뷰에서 타입 생성시 어떻게 구현할지 고민해보라는 말씀을 듣게 되었고, 말씀 중에 `런타임에 실행하지 않고 컴파일 타임에 점검할 수 있으면 가장 좋다` 는 말씀을 어떻게 적용해볼지 고민했습니다. 
  저는 failable init을 사용하면 해당 사항을 충족시킬 것이라 판단하여 Alpha, BackgroundColor 타입의 초기화 생성자로 failable한 생성자를 사용하도록 수정했습니다.

- 사각형 뷰와 Rectangle 객체의 매칭을 위해 어떤 방식을 사용할 수 있을까 고민하다가, Rectangle 객체의 id 프로퍼티는 고유한 값이 되기 때문에 사각형 뷰를 RectangleView로 선언해주고, UIView를 상속받도록 했습니다. 그리고 새로운 속성인 id를 가질 수 있도록 설정했습니다.

- Alpha를 OpacityLevel이라는 매개변수로 초기화를 시켜줬었는데, 편하게 코드를 작성하려고 하다보니 해당 값을 그냥 UIView 등에서 사용하는 alpha와 유사한 Double으로 초기화해주는 초기화 생성자를 하나 더 만들어줬습니다. 그런데 Double로 초기화를 하다보니, 초기화 시 줄 수 있는 값을 제한하는게 불가능하다고 생각되어 다시 해당 초기화 생성자를 제거했습니다.

- UITabGestureRecognizer를 이용하는 방법을 공부하다가, hitTest 메서드를 이용하면 터치한 좌표의 최상단 뷰를 가져올 수 있다는걸 배웠습니다. 그런데 다른 팀원들 리뷰를 보다가 터치 이벤트 흐름을 VC에서는 터치한 좌표를 Plane에게 알려만 주고, Plane이 해당 좌표에 존재하는 Rectangle을 반환하면 그에 맞게 테두리를 그리는 식으로 설정해보라는 말씀을 보게 됐습니다.
  저는 이 흐름을 만들기 위해 어떤 방식을 사용할까 고민하다가, Delegate 패턴을 사용하는 방식으로 수정해보았습니다.

Hansolkkim committed:
Style: RectangleFactory 클래스의 generateRandomRectangle(in:) 메서드 튜플 형태 매개변수에 label 명시

Hansolkkim committed:
Refactor: 사각형 View에 해당하는 RectangleView 구현

사각형 뷰에 해당하는 UIView가 id 프로퍼티를 가질 수 있도록 RectangleView 구현
Rectangle 객체의 id 프로퍼티 접근제어자 private(set) 으로 수정

Hansolkkim committed:
Feat: drawableAreaView에 GestureRecognizer 추가

뷰가 터치되면 테두리에 선을 표시해 선택된것을 알 수 있도록 구현

Hansolkkim committed:
Feat: 터치된 RectangleView에 맞춰 backgroundButton의 title, backgroundColor 속성이 변경

backgroundButton의 아울렛변수 선언
updateBackgroundButton(with:) 메서드를 이용해 backgroundButton의 속성 변경
backgroundButton의  title 속성에 나타내는 String을 만들기 위해 BackgroundColor 타입에 hexCode 프로퍼티 추가
Plane 클래스에서 터치된 RectangleView에 해당하는 Rectangle 객체를 찾아 반환해주는 getRectangle(id:) 메서드 선언
getRectangle(id:) 메서드에서 ID 비교를 위해 ID 타입 구조체가 Equtable 프로토콜 채택

Hansolkkim committed:
Feat: 터치된 RectangleView에 맞춰 alphaSlider의 헤드가 이동하는 기능 추가

alphaSlider의 아울렛변수 선언
updateAlphaSlider(with:) 메서드를 이용해 alphaSlider의 헤드 이동

Hansolkkim committed:
Refactor: ViewController의 아울렛변수 업데이트 메서드 매개변수 수정

해당 아울렛변수 업데이트 메서드에서 필요한 매개변수만 받아오도록 매개변수의 타입, 이름 수정

Hansolkkim committed:
Style: PlaneDelegate 프로토콜 이름 GenerateRectangleViewDelegate로 변경

Hansolkkim committed:
Style: Plane 클래스의 getRectangle(id:) 메서드 이름 specifyRectangle(id:) 로 변경 & 선택된 View 업데이트하는 부분 메서드 분리

선택된 View를 업데이트하는 부분을 updateSelectedView(_:) 메서드로 분리

Hansolkkim committed:
Remove: 중복 추가된 파일과 사용하지 않는 메서드 제거

Hansolkkim committed:
Refactor: Alpha, BackgroundColor 타입의 초기화 생성자를 failable한 생성자로 변경

해당 타입 사용 시에 잘못된 값을 입력하면 만들어지지 않도록 failable init 사용

Hansolkkim committed:
Style: BackgroundColor 생성자 내의 colors 상수 명 colorValues로 수정

Hansolkkim committed:
Style: Alpha,Background,Point를 랜덤 생성해주는 메서드를 따로 분리해 AlphaFactory, BackgroundColorFactory, PointFactory 객체로 분리

backgroundButton 클릭시 배경색을 랜덤 변환하는 코드를 재사용하기 위해 분리

Hansolkkim committed:
Style: ViewController의 backgroundButton 변수명을 backgroundColorButton 으로 수정

모호한 이름 수정

Hansolkkim committed:
Feat: 앱의 backgroundColor 버튼을 누르면 선택된 View의 backgroundColor를 랜덤하게 바꿔주는 기능 추가

ViewController에서 선택된 View를 프로퍼티로 가지도록 touchedView 프로퍼티 선언
Rectangle 객체의 backgroundColor를 랜덤하게 변화시키는 changeBackgroundColorRandomly() 메서드 구현
ViewController로부터 특정 View의 색깔을 바꾸라는 지시를 받으면 Plane에서 해당하는 Rectangle의 색을 바꿔주는 changeBackgroundColorOfRectanlge(id:) 메서드 추가
Rectangle 객체의 배경 색이 바뀌면 특정 View의 색깔도 바뀌는 Delegate 패턴을 위한 ChangeRectangleBackgroundColorDelegate 프로토콜 선언
ViewController가 해당 역할을 위임받아 View의 색깔을 바꿀 수 있도록 하는 rectangleDidChangeBackgroundColor(to: alpha:) 메서드 구현

Hansolkkim committed:
Refactor: 불필요한 Delegate 패턴 제거

ViewController에서 backgroundButton이 터치되면 선택된 뷰의 색을 임의대로 바꾸고 Plane에게 같은 색으로 해당 Rectangle의 BackgroundColor를 바꾸도록 지시하는 순서로 수정
그 과정에서 불필요했던 Delegate 패턴 제거
Plane 클래스와 Rectangle 클래스에서 Rectangle의 바뀔 BackgroundColor를 매개변수로 넘겨줄 수 있도록 메서드의 매개변수 추가
ChangeRectangleBackgroundColorDelegate.swift 파일 제거 및 해당 파일에 있던 프로토콜 제거

Hansolkkim committed:
Feat: 앱의 alphaSlider의 value를 변경해주면 View의 alpha를 변경된 값으로 바꿔주는 기능 추가

Rectangle 객체의 alpha를 바꿔주는 기능을 하는 changeAlphaValue(to:) 메서드 추가
ViewController로부터 특정 View의 색깔을 바꾸라는 지시를 받으면 Plane에서 해당하는 Rectangle의 alpha를 바꿔주는 changeAlphaValueOfRectanlge(id:to:) 메서드 추가
Alpha 타입에 원하는 Double 값으로 value를 초기화 시킬 수 있는 failable 초기화 생성자 추가
스토리보드에서 alphaSlider에 해당하는 UISlider의 minimum Value를 0.1로 설정

Hansolkkim committed:
Feat: 사각형 뷰가 없는 빈 공간을 터치하면 기존의 사각형 선택이 취소되는 기능 추가

배경색 버튼의 .disabled 상태일 때의 title 선언
빈 공간을 터치한 경우도 오류가 아닌 의미있는 경우로 사용하기 위해 viewDidTouched(sender:) 메서드에서 hitTest(_:with:) 메서드를 이용할 때 guard let 구문 대신 if let 구문 사용
빈공간을 터치했을 때의 경우에 발생할 이벤트를 if let의 else 구문에 추가
updateBackgroundButton(color:alpha:), updateAlphaSlider(alpha:) 메서드에 해당 버튼/슬라이더를 활성화하는 코드 추가

Hansolkkim committed:
Refactor: GestureRecognizer에서의 동작 흐름 변경

GestureRecognizer에서의 동작을 "어떤 부분을 터치하면 해당 부분의 좌표를 Plane에게 넘겨주는 흐름" 으로 변경
해당 흐름을 위한 UpdateViewMatchedRectangleDelegate 프로토콜 선언
해당 흐름을 위해 터치된 좌표로 Rectangle객체를 특정할 수 있게하도록 Plane 클래스의 specifyRectangle(id:) 메서드의 매개변수를 ID 타입으로 변경
(사각형이 generate되는 구조상, 더 위에 있는 사각형은 나중에 생성된 사각형이므로 specifyRectangle(id:) 메서드에서 rectangles 배열 프로퍼티를 reversed하여 해당 rectangle을 찾기위해 순회하도록 설정)
Rectangle 클래스에 좌표를 받으면 해당 좌표가 본인의 영역에 포함되는지 여부를 반환하는 기능을 하는 isPointInArea(_:) 메서드 추가
ViewController에서 UpdateViewMatchedRectangleDelegate 프로토콜을 채택하여, 터치된 좌표로 사각형이 특정되면 처리해야할 내용을 rectangleDidSpecified(_:) 메서드에서 처리
특정된 Rectangle로 인해 View를 업데이트하는 private 메서드들을 해당 부분과 관련있는 위치로 이동
빈공간을 터치했을 경우를 처리하기 위한 initializeViewsInTouchedEmptySpaceCondition() 메서드를 ViewController 클래스 내에 선언

Hansolkkim committed:
Feat: alphaSlider로 인해 View의 alpha가 변하면 해당 값에 맞게 backgroundColorButton의 backgroundColor의 alpha도 변하는 기능 추가

Hansolkkim committed:
Feat: minus/plusAlphaValueButton을 이용해 선택된 View의 Alpha를 줄여주고 늘려주는 기능 추가

해당 버튼들을 조건에 맞게 업데이트 시키는 메서드를 선언하고, 특정 위치에서 새로운 alpha값으로 StatusView에 있는 요소 View들을 업데이트시켜줄 수 있도록 updateStatusViewElement(with:) 메서드 구현

Hansolkkim committed:
Fix: Alpha 초기화 시 발생하는 오류를 위한 수정

Alpha를 Double 타입으로 초기화하는 생성자를 추가했는데, 해당 생성자를 이용하면 오류가 많이 발생하여 해당 생성자 제거

Hansolkkim committed:
Style: 0.55 하드코딩 부분 의미있는 상수로 선언 및 오타 수정

Hansolkkim committed:
Refactor: 지속적으로 사용되는 변환 코드를 새로 생성한 Converter 구조체에서 static 함수인 convertToUIColor(backgroundColor:) 로 생성

Hansolkkim committed:
Test: Plane 객체 테스트를 위한 유닛테스트 추가

테스트를 위해 Plane 클래스의 메서드들 중 반환값이 없었던 일부 메서드에 반환값이 있도록 변경
테스트를 위한 addSpecificRectangle(_:) 메서드를 Plane 클래스에 추가
테스트를 위해 Size, BackgroundColor, Alpha 타입 구조체들과 Rectangle 구조체가 Equatable 프로토콜을 채택

Hansolkkim committed:
Fix: Test 결과 발견한 문제점 수정

Plane에 특정 Rectangle 객체의 Alpha를 변경하도록 지시할 때 CGFloat 타입이 아닌 Alpha 타입을 매개변수로 넘겨주도록 수정
Test로 Plane이 특정 Rectangle의 backgroundColor를 바꾸려고 할 때 이중으로 랜덤 BackgroundColor 타입을 생성받아 사용하는 과정을 찾아, 해당 부분 수정

Hansolkkim committed:
Refactor: Plane 클래스를 구조체로 변경

Plane 객체의 기능상 mutating function은 Rectangle 객체를 추가할 때 뿐이고, 큰 값을 구조체 내의 속성에 저장하지 않기 때문에 굳이 class일 필요가 없다 판단하여 class -> struct로 변환

Hansolkkim committed:
Refactor: Rectangle의 값이 변하는 위치에서 delegate의 메서드를 사용하도록 수정

Hansolkkim committed:
Refactor: Rectangle 클래스의 ==(lhs:rhs:) 메서드 수정

id만 같으면 무조건 같은 사각형이기 때문에 부수적인 비교문 제거

Hansolkkim committed:
Fix: 배경색 버튼을 누르면 배경색 버튼의 alpha가 1.0이 되는 문제 수정

사용했던 상수 previousBackgroundButtonAlpha가 수정 후 어울리지 않아, previousAlpha로 수정

godrm commented on DrawingApp/DrawingApp/Model/Alpha.swift:
```diff
@@ -7,18 +7,19 @@
     let value: Double
     static let possibleOpacityLevels = 1...10
+    static let possibleAlphaValues = stride(from: 0.1, to: 1.0, by: 0.01)
     
-    init(opacityLevel: Int) {
+    init?(opacityLevel: Int) {
         let min = Alpha.possibleOpacityLevels.min() ?? 1
         let max = Alpha.possibleOpacityLevels.max() ?? 10
         
         if opacityLevel > max {
-            self.value = 1.0
+            return nil
         } else if opacityLevel < min {
-            self.value = 0.1
+            return nil
```
opacityLevel를 범위로 걸러내는 것도 좋은 방식입니다만 매개변수가 Int 니까 컴파일 타임에 걸러지지는 않겠죠
컴파일 타임에 확인되려면 opacityLevel 레벨이 예를 들어 1...10까지만 들어갈 수 있는 enum 타입이거나 별도 타입이어야 겠죠
그리고 단계를 정수로 주는데 value가 double이 되는 것도 살짝 어색합니다. 그 관계를 명확하게 알기 어려우니까요
지정한 값을 확인할 수 있고, doubleValue를 가져갈 수도 있는 구조라면 더 좋을 것 같습니다.

Hansolkkim replied:
> opacityLevel를 범위로 걸러내는 것도 좋은 방식입니다만 매개변수가 Int 니까 컴파일 타임에 걸러지지는 않겠죠
> 컴파일 타임에 확인되려면 opacityLevel 레벨이 예를 들어 1...10까지만 들어갈 수 있는 enum 타입이거나 별도 타입이어야 겠죠
> 그리고 단계를 정수로 주는데 value가 double이 되는 것도 살짝 어색합니다. 그 관계를 명확하게 알기 어려우니까요
> 지정한 값을 확인할 수 있고, doubleValue를 가져갈 수도 있는 구조라면 더 좋을 것 같습니다.

어제 OCP 강의 때 사용하셨던 방법 이용해서 OpacityLevel이 될 수 있는 프로토타입들을 단계(1...10)만큼 만들어 사용하는 방식으로 수정해보겠습니다.

Hansolkkim replied:
@godrm 이 부분에 대해서 질문이 있습니다! 
Alpha에 들어갈 수 있는 값을 제한해보라고 하셨는데, 제한하는 방법을 생각해봐도 UISlider.value와 같은 값을 이용하기 위해서는 결국은 Alpha든, OpacityLevel 타입이든 CGFloat와 같은 원시값으로 초기화하는 과정이 필요하게 돼서 여기서 헤메는 중입니다..
기존에 RectangleFactory 등에서는 Alpha에 들어가는 값을 컴파일 타임에서 확인할 수 있도록 제한하는 방법은 이해가 갔는데,
UISlider.value로 Alpha 타입을 만들어줄 때의 값을 제한하는 방법은 이해가 가지 않습니다ㅠㅠ

godrm replied:
Alpha에서 할 일과 상위에서 변환해서 넘겨줄 일을 구분해야하는 거죠 (두 개 피드백이 합쳐져 있었네요)
우선 매개변수 Int 랑 내부 속성 Double 타입인 것은 Alpha 내부 사정이지만 이것을 맞춰보시구요

---
UISlider.value로 만들어 줄때는 외부 입장에서 Alpha에 맞춰야 하는 상황이죠. 
제 말은 Alpha부터 만들면 외부가 복잡해질테니, 필요하면 Alpha에 다른 생성자가 추가될 수도 있겠죠
그렇지만 UISlider.value가 아니더라도 Alpha 자체만 놓고 봐도, 단위 테스트만 하더라도 충분히 안정적으로 생성되고 확인할 수 있어야 합니다. 예를 들어 10으로 넣고 생성했는데 value가 1.0라면 예측가능하지 않아서 테스트하기 어려울 것 같다는 맥락입니다. 

---
변환하는 것은 위의 것과 별개로 상위에서 CGFloat -> Int로 바꾸거나 CGFloat -> Alpha로 바꾸는 함수가 필요할 수 있습니다. 그건 여기 init을 고치라는 의미가 아닙니다. CGFloat 과 Alpha를 둘 다 알고 있는 상위에서 변환하는 기능이 필요한거죠

Hansolkkim replied:
Alpha의 OpacityLevel에 해당하는 열거형을 선언해주어 초기화에 사용하도록 하여 컴파일 타임에서 입력 가능한 범위를 제한해보고, Alpha의 프로퍼티에 value 외에 opacityLevel도 추가하여 생성 시점과 확인 시점에서의 어색함을 줄여보겠습니다!

제가 초기화 시점에서의 범위 제한과 타입 생성에 대해서 복잡하게 생각하다보니 많이 꼬여버린 것 같습니다. 답변 감사합니다!

Hansolkkim replied:
Alpha의 OpacityLevel에 해당하는 열거형을 선언해주어 초기화에 사용하도록 하여 컴파일 타임에서 입력 가능한 범위를 제한해보고, Alpha의 프로퍼티에 value 외에 opacityLevel도 추가하여 생성 시점과 확인 시점에서의 어색함을 줄여보겠습니다!

제가 초기화 시점에서의 범위 제한과 타입 생성에 대해서 복잡하게 생각하다보니 많이 꼬여버린 것 같습니다. 답변 감사합니다!

godrm commented on DrawingApp/DrawingApp/Model/AlphaFactory.swift:
```diff
@@ -0,0 +1,15 @@
+import Foundation
+
+class AlphaFactory {
+    public static func generateRandomAlpha() -> Alpha? {
```
스위프트 API 디자인 가이드상에서 값을 생성하는 경우는 make-를 주로 사용합니다
그리고 팩토리는 특정한 값으로 안전하게 만들어지는 것을 보장해주기 때문에 옵셔널이 아니어도 좋습니다. 
그런게 Alpha.init()과 다른 점이죠

Hansolkkim replied:
> 스위프트 API 디자인 가이드상에서 값을 생성하는 경우는 make-를 주로 사용합니다
> 그리고 팩토리는 특정한 값으로 안전하게 만들어지는 것을 보장해주기 때문에 옵셔널이 아니어도 좋습니다.
> 그런게 Alpha.init()과 다른 점이죠

컨벤션이 있는줄 몰랐네요,, 팩토리 메서드들의 이름 `make-` 형태로 수정하고, 팩토리 메서드들의 반환값 타입이 옵셔널이 아니도록 수정해보겠습니다.

Hansolkkim replied:
> 스위프트 API 디자인 가이드상에서 값을 생성하는 경우는 make-를 주로 사용합니다
> 그리고 팩토리는 특정한 값으로 안전하게 만들어지는 것을 보장해주기 때문에 옵셔널이 아니어도 좋습니다.
> 그런게 Alpha.init()과 다른 점이죠

컨벤션이 있는줄 몰랐네요,, 팩토리 메서드들의 이름 `make-` 형태로 수정하고, 팩토리 메서드들의 반환값 타입이 옵셔널이 아니도록 수정해보겠습니다.

godrm commented on DrawingApp/DrawingApp/Model/BackgroundColor.swift:
```diff
@@ -7,27 +7,33 @@
     
-    init(r: Int, g: Int, b: Int) {
-        var colors = [Double(r), Double(g), Double(b)]
+    init?(r: Int, g: Int, b: Int) {
+        let colorValues = [Double(r), Double(g), Double(b)]
         let min = BackgroundColor.possibleColorValues.min() ?? 0
         let max = BackgroundColor.possibleColorValues.max() ?? 255
```
Alpha와 마찬가지인데 BackgroundColor도 init 할 때는 Int로 정수로 입력하는 데 생성된 값은 Double이라 어떤 관계가 있는지 유추하기 어려울 것 같습니다. 
구현을 모르는 테스트를 한다고 생각해보면 어떤 값을 넣어서 어떤 값이 생성되는지 알 수 있을까요? 
생성 시점 매개변수 타입과 속성의 타입이 다르면 좀 혼란스러울 것 같습니다

Hansolkkim replied:
> Alpha와 마찬가지인데 BackgroundColor도 init 할 때는 Int로 정수로 입력하는 데 생성된 값은 Double이라 어떤 관계가 있는지 유추하기 어려울 것 같습니다.
> 구현을 모르는 테스트를 한다고 생각해보면 어떤 값을 넣어서 어떤 값이 생성되는지 알 수 있을까요?
> 생성 시점 매개변수 타입과 속성의 타입이 다르면 좀 혼란스러울 것 같습니다

저도 저 부분을 왜 저렇게 바꿨는지 의아하네요...ㅎㅎㅎ..
해당 부분 Int로 수정하겠습니다..

godrm commented on DrawingApp/DrawingApp/Model/Converter.swift:
```diff
@@ -0,0 +1,22 @@
+import UIKit
+
+struct Converter {
+    public static func convertToUIColor(backgroundColor: BackgroundColor, alpha: Double = 1.0) -> UIColor {
```
별도 타입을 만들기 보다는 관련있는 타입을 어느 파일에서 확장해서 쓰면 어떨까 생각해보세요
Converter는 오히려 불필요한 타입입니다

Hansolkkim replied:
> 별도 타입을 만들기 보다는 관련있는 타입을 어느 파일에서 확장해서 쓰면 어떨까 생각해보세요
> Converter는 오히려 불필요한 타입입니다

파일에서의 특정 타입을 확장하여 수정해보겠습니다! 안그래도 해당 부분을 작성하고 나니 JK께서 이전에 말한 유틸리티에 해당하는 걸 작성한게 아닌가 싶었는데 맞았네요. .감사합니다!

godrm commented on DrawingApp/DrawingApp/Model/GenerateRectangleViewDelegate.swift:
```diff
@@ -7,6 +7,6 @@
 
-protocol PlaneDelegate {
+protocol GenerateRectangleViewDelegate {
     func planeDidAddRectangle(_ rectangle: Rectangle)
```
델리게이트 수업할 때 언급한 것처럼 이름과 매개변수 규칙을 지켜보세요

Hansolkkim replied:
> 델리게이트 수업할 때 언급한 것처럼 이름과 매개변수 규칙을 지켜보세요

델리게이트 메서드의 이름 첫 자리는 이벤트의 주체가 되도록, 그리고 매개변수와 같은 타입이도록 해당 메서드 이름을 `rectangleDidAdded(_ rectangle: Rectangle)` 로 수정해보겠습니다!

godrm commented on DrawingApp/DrawingApp/Model/Plane.swift:
```diff
@@ -18,15 +19,45 @@ class Plane {
+                updateViewMatchedRectangleDelegate?.rectangleDidSpecified(rectangle)
+                return rectangle
+            }
+        }
```
혹시 사각형을 찾지 못한 경우는 위로 알려줄 필요가 없을까요?

Hansolkkim replied:
> 혹시 사각형을 찾지 못한 경우는 위로 알려줄 필요가 없을까요?

알려주기 위해, rectangleDidSpecified(_:) 메서드의 매개변수가 Optional 타입을 받고, 사각형을 찾지 못 한 경우 해당 메서드의 매개변수를 nil로 하여 호출하도록 설정하여 수정해보겠습니다!

godrm commented on DrawingApp/DrawingApp/Model/Plane.swift:
```diff
@@ -18,15 +19,45 @@ class Plane {
+        return nil
+    }
+    
+    mutating public func addSpecificRectangle(_ rectangle: Rectangle) { // 테스트용 메서드
+        self.rectangles.append(rectangle)
     }
```
테스트용 메소드라는 게 어떤 의미인거죠? 이 메소드는 실제 사용하는게 아니고 테스트할 때만 쓰는건가요?
테스트할 때만 사용하는 메소드라는 게 존재한다면 이상한 것 같습니다.

Hansolkkim replied:
> 테스트용 메소드라는 게 어떤 의미인거죠? 이 메소드는 실제 사용하는게 아니고 테스트할 때만 쓰는건가요?
> 테스트할 때만 사용하는 메소드라는 게 존재한다면 이상한 것 같습니다.

그런 이유로 사용했습니다.. 해당 부분 수정하고 테스트 코드 수정해보겠습니다!

godrm commented on DrawingApp/DrawingApp/Model/Plane.swift:
```diff
@@ -18,15 +19,45 @@ class Plane {
+    }
+    
+    public func changeBackGroundColorOfRectangle(id: ID, to newColor: BackgroundColor) -> Rectangle? {
+        for rectangle in rectangles {
+            if rectangle.id == id {
+                rectangle.changeBackgroundColor(to: newColor)
+                updateViewMatchedRectangleDelegate?.rectangleBackgroundColorDidChanged(rectangle)
+                return rectangle
+            }
+        }
```
이런 경우 상위에서 ID를 계속 전달해줘야 하는데 
선택된 사각형 정보를 뷰 컨트롤러가 갖고 있는게 좋을지, 여기 Plane에서 처리하는 게 좋을지도 고민해보세요
누구의 역할이고 누구의 책임이 좋을까요? 뷰 컨트롤러는 어떤 역할을 주로 하던가요?

Hansolkkim replied:
이 부분에서는 뷰 컨트롤러가 뷰로 부터 변화를 감지하고, 모델(Plane)에게 해당 변화로 인한 업데이트를 요구하는 역할을 합니다.
해당 메서드는 뷰 컨트롤러로부터 Plane이 특정 Rectangle의 backgroundColor를 변경하라는 지시를 받아서 변경하는 메서드인데, 이런 경우에는 뷰 컨트롤러가 Plane에게 Rectangle을 특정해서 넘겨주고 바꾸라고 지시하는게 맞는 흐름일까요?

일단, 아래의 리뷰들을 참고해서 id 프로퍼티로 Rectangle을 비교하지 않기 위해 뷰 컨트롤러에서 RectangleView와 Rectangle을 매핑해놓는 프로퍼티를 만들어 소유하고 있도록하고, 위의 메서드를 호출할 때 Rectangle을 특정해서 매개변수로 넘겨주는 방식으로 수정해보려고 합니다. 그런데 이렇게 뷰 컨트롤러가 특정 Rectangle까지 알고 있는게 맞는가 궁금합니다.

Hansolkkim replied:
해당 부분, Plane에서는 이전에 선택된 Rectangle을 알고 있도록 하는 프로퍼티를 만들어 알고있도록 하고, 뷰컨트롤러로부터 변화(backgroundColor/ alpha) 지시를 받은 경우에 그 프로퍼티에 해당하는 Rectangle을 변화하도록 수정했습니다.
그리고 앞서 말씀드린 매핑해놓은 프로퍼티의 경우에는 Key가 Rectangle이고, Value가 RectangleView인 Dictionary인데, 뷰컨트롤러의 프로퍼티가 하위모델인 Rectangle을 직접 안다기보다는, 어떤게 뷰와 매칭되는 Rectangle 객체인지 정도만 아는 것이라 판단되어 이를 계속 사용하도록 하려고 합니다.

godrm commented on DrawingApp/DrawingApp/Model/Rectangle.swift:
```diff
@@ -8,7 +8,7 @@
 
 class Rectangle {
-    private let id: ID
+    private(set) var id: ID
```
id를 속성으로 가져가지 않고 Equatable을 채택했으면 Rectangle 자체를 비교할 수 있지 않을까요?

Hansolkkim replied:
id를 유의미하게 사용해보고자 하는 생각으로 id를 속성으로 가지고 가서 비교하는 방법을 사용했습니다. 그런데 이렇게 내부의 속성을 사용하지 않고도 비교할 수 있다면 그런 방법을 사용하는게 더 적절하다고 생각됩니다.
해당 부분을 다시 `private let id: ID` 로 수정하도록 하고, 비교에서는 JK 말씀처럼 Rectangle 자체를 비교하는 방법을 사용해보겠습니다.

godrm commented on DrawingApp/DrawingApp/Model/Rectangle.swift:
```diff
@@ -30,8 +30,20 @@ class Rectangle {
-    public func hasSamePoint(_ point: Point) -> Bool {
-        return self.point == point
+    func isPointInArea(_ point: Point) -> Bool {
+        if point.x >= self.point.x && point.x <= self.point.x + self.size.width &&
+            point.y >= self.point.y && point.y <= self.point.y + self.size.height {
+            return true
+        }
+        return false
```
이런 경우 return `조건식` 형태로 사용해도 깔끔합니다. 참고하세요

Hansolkkim replied:
이전에도 받았던 리뷰인데 생각처럼 잘 못 쓰게 되네요.. 수정하도록 하겠습니다!

godrm commented on DrawingApp/DrawingApp/View/RectangleView.swift:
```diff
@@ -0,0 +1,33 @@
+import UIKit
+
+class RectangleView: UIView {
+    let id: ID
+    override var description: String {
+        return "ID = \(id); Frame = \(frame); alpha = \(alpha); backgroundColor = \(backgroundColor ?? .white)"
+    }
+    
+    required override init(frame: CGRect) {
+        self.id = ID.init()
```
뷰를 생성하는 데 ID가 꼭 필요한 것일까요?
뷰가 데이터를 저장하고 그걸 가져가서 또 비교하는 것보다는 뷰도 뷰의 인스턴스 자체를 비교하는 방식이 좋습니다.
뷰와 모델을 매칭할 때도 ID가 아니라 인스턴스끼리 비교하도록 해보세요

Hansolkkim replied:
넵 RectangleView 객체에서 id 프로퍼티를 제거하고 인스턴스끼리 비교하는 방법을 사용하도록 수정해보겠습니다!

godrm commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -10,17 +10,29 @@ import OSLog
 class ViewController: UIViewController {
     
-    let plane = Plane()
+    var plane = Plane()
```
ViewController의 속성도 private var를 우선적으로 고려하세요

Hansolkkim replied:
제일 상위 모듈으로만 사용할거라고 생각해 그 부분을 신경 못 썼던 것 같습니다. 프로퍼티들을 private 접근 제어자를 사용하다록 수정하겠습니다.

godrm commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -53,16 +65,201 @@ class ViewController: UIViewController {
+    }
+    
+    private func initializeViewsInTouchedEmptySpaceCondition() {
+        self.touchedView?.layer.borderWidth = 0
+        self.touchedView = nil
+        backgroundColorButton.isEnabled = false
+        backgroundColorButton.backgroundColor = .clear
+        alphaSlider.isEnabled = false
+        let alphaSliderCenterValue = (alphaSlider.minimumValue + alphaSlider.maximumValue) / 2
+        alphaSlider.setValue(alphaSliderCenterValue, animated: true)
+        minusAlphaValueButton.isEnabled = false
+        minusAlphaValueButton.backgroundColor = .systemGray6
+        plusAlphaValueButton.isEnabled = false
+        plusAlphaValueButton.backgroundColor = .systemGray6
```
이런 뷰를 바꾸는 동작이 출력에 해당합니다. 터치했을 때 바로 이걸 변경하지 마시고, 
출력의 흐름을 만들고 모델에서 뷰컨트롤러로 알려줄 때 출력하도록 해보세요

Hansolkkim replied:
> 알려주기 위해, rectangleDidSpecified(_:) 메서드의 매개변수가 Optional 타입을 받고, 사각형을 찾지 못 한 경우 해당 메서드의 매개변수를 nil로 하여 호출하도록 설정하여 수정해보겠습니다!

위에서 말씀드렸던 방법을 이용해 모델에서 뷰 컨트롤러로 알려줄 때인 rectangleDidAdded(_:) 메서드에서 매개변수에 nil을 사용해 호출하고, 해당 매개변수가 nil일 경우에 해당 메서드를 호출하도록 수정해보겠습니다.

godrm commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -53,16 +65,201 @@ class ViewController: UIViewController {
+    }
+    
+    @IBAction func alphaSliderValueChanged(_ sender: UISlider) {
+        let newAlphaValue = Double(String(format: "%.1f", sender.value)) ?? 0
+        let convertedOpacityLevel = Int(10 * newAlphaValue)
```
여기 값을 다루는 게 이상하게 복잡해보이네요. sender.value가 어떤 값이길래 String으로 바꿨다가 다시 Double로 바꾸는건가요?
더구나 다시 Int로 또 바꾸는 과정이 번거로워 보입니다. 개선할 수 없을까요?

Hansolkkim replied:
원하는 길이로 자른 다음 Alpha를 초기화하는데 사용하려고 위와 같이 사용했습니다..
Alpha를 초기화하기 위한 값의 타입을 위에서 바꿨기 때문에 Int로 바꾸는 부분은 삭제했고,
newAlphaValue를 구하는 부분을 타입 변환을 계속해서 사용하는 방법 대신 `Double(round(sender.value * 10) / 10.0)` 으로 수정해보겠습니다!

Hansolkkim replied:
해당 부분 수정하다가 `sender.value` 값이 Float 타입인데, Double 타입으로 형변환하는 과정에서 불필요한 소수값이 추가되는 문제가 발생해, Alpha 등에서 사용하는 타입을 모두 Float으로 수정하고, 위에서 말씀드린 방식을 `round(sender.value * 10) / 10` 으로 변경하여 적용하겠습니다.

godrm commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -53,16 +65,201 @@ class ViewController: UIViewController {
+        plusAlphaValueButton.isEnabled = true
+        plusAlphaValueButton.backgroundColor = .white
+        
+        let newAlphaValue = Double(String(format: "%.1f",(touchedView.alpha + 0.1))) ?? 1.0
+        let convertedOpacityLevel = Int(10 * newAlphaValue)
```
위에 언급한 것처럼 변환 과정이 번거롭고 계속 반복됩니다.

Hansolkkim replied:
해당 부분도 위에서 답변 드린 대로, `round((touchedView.alpha + 0.1) * 10) / 10.0` 로 바꾸도록 해보겠습니다.

godrm commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -53,16 +65,201 @@ class ViewController: UIViewController {
+        guard let matchedView = findMatchedView(has: specifiedRectangle.id) else {
+            return
+        }
+        self.touchedView = matchedView
+        
+        if let specifiedRectangleView = self.touchedView {
```
self.touchedView 에 넣고 또 specifiedRectangleView 바인딩하는 게 좀 어색합니다. 
matchedView만으로 가능한 게 아닐까요?

Hansolkkim replied:
아래의 바인딩부분 제거하고 matchedView만 이용하는 방법으로 수정하겠습니다!

godrm commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -53,16 +65,201 @@ class ViewController: UIViewController {
+    }
+    
+    private func findMatchedView(has id: ID) -> RectangleView? {
+        for rectangleView in drawableAreaView.subviews {
+            let rectangleView = rectangleView as? RectangleView
+            if rectangleView?.id == id {
+                return rectangleView
+            }
+        }
```
서브뷰에서 O(n) 방식으로 다 찾기보다는 매칭해서 O(1)으로 찾는 게 더 좋습니다.

Hansolkkim replied:
위에서 답변 드린 내용으로, RectangleView와 Rectangle을 viewRectangleMap 이라는 `Dictionary<RectangleView: Rectangle>` 프로퍼티에 생성시 매칭해서 넣어주는 방법으로 수정했습니다. 그런데 Delegate 메서드에서 Rectangle 객체를 매개변수로 받아와서 사용하다보니, 다시 View를 찾기 위한 방법은 O(1) 으로 사용하기가 어려워졌습니다.. 
```swift
let matchedView = viewRectangleMap.first(where: { view, rectangle in
    rectangle == specifiedRectangle
})?.key
```
혹시 해당 방법 외에 Value로 Key를 접근할 때 O(1)의 시간복잡도를 갖도록 하는 방법이 있을까요? 아니면 제가 매칭한 방법에서는 O(1)의 방법이 없는걸까요?

Hansolkkim replied:
해당 부분을 지우고, ViewController가 Rectangle과 RectangleView를 매칭해서 알고있도록 rectangleAndViewMap Dictionary 타입 프로퍼티를 사용하고, 변경된 Rectangle 객체를 델리게이트 메서드의 매개변수로 가져와서 해당 프로퍼티의 Key값으로 사용해 O(1)으로 RectangleView를 찾을 수 있도록 수정하겠습니다!

godrm reviewed (changes requested):
입력과 출력 흐름을 나누고, 뷰를 만들 때 반복적으로 값이 복잡하게 처리되는 데 이 부분을 개선해보세요

Hansolkkim committed:
Refactor: Alpha의 OpacityLevel에 대한 열거형 지정

Alpha의 초기화 메서드 수정

Hansolkkim committed:
Refactor: Factory 관련 수정

컨벤션에 맞게 Factory 메소드들의 이름을 make- 로 변경
Factory 메소드들이 Optional 값이 아닌 값을 반환하도록 수정

Hansolkkim committed:
Fix: BackgroundColor의 생성 시점 매개변수 타입과 속성의 타입 일치

Hansolkkim committed:
Refactor: Converter 타입 제거 및 backgroundColor를 UIColor로 변환해주는 메서드를 fileprivate 접근 제어자로 제한해 BackgroundColor 타입 확장

Hansolkkim committed:
Refactor: GenerateRectangleDelegate 프로토콜 수정

Delegate명 수정
Delegate 이름 규칙에 맞게 메서드 이름 수정
Plane과 연관된 delegate 메서드들을 하나의 프로토콜로 통합

Hansolkkim committed:
Refactor: RectangleView에 id 프로퍼티 제거

RectangleView의 id를 Rectangle과 매치하지 않고, ViewController에 viewRectangleMap 이라는 Dictionary<RectangleView: Rectangle> 프로퍼티를 만들어 매칭되는 RectangleView와 Rectangle을 ViewController가 알고 있도록 설정
그에 따른 메서드 내용 수정

Hansolkkim committed:
Refactor: 비교문 간단한 표현으로 수정

Hansolkkim committed:
Refactor: UISlider의 value 속성 변환식 수정

UISlider의 value 속성을 원하는 값으로 변환하기 위한 수식을 좀 더 간단하게 수정

Hansolkkim committed:
Style: 들여쓰기 잘못된 부분 수정

Hansolkkim committed:
Refactor: viewRectangleMap 프로퍼티 변경

해당 프로퍼티의 이름을 변화된 타입에 맞게 rectangleAndViewMap으로 변경
ViewController에서 변경된 Rectangle과 일치하는 RectangleView를 찾을 때의 시간 복잡도를 O(1)로 만들기 위해 rectangleAndViewMap의  프로퍼티의 Key 타입을 Rectangle로 변경
Rectangle 객체를 rectangleAndViewMap의 Key로 사용하기 위해 Rectangle 클래스가 Hashable 프로토콜 채택하도록, id 구조체가 Hashable 프로토콜을 채택하도록 설정
Plane 구조체에서 변경해야할 Rectangle을 알고있도록 하기 위해 Plane 구조체에 specifiedRectangle 프로퍼티 추가
Alpha를 변경하는 버튼에 맞게 동작을 하도록 Plane 구조체에 stepUpAlphaValue(), stepDownAlphaValue() 메서드 선언

Hansolkkim committed:
Refactor: Alpha의 value 프로퍼티의 타입 Double -> Float 변경

ViewController의 alphaSliderValueChanged(_:) 액션 메서드에서 변경될 alpha값을 변환하는 과정을 줄이기 위해 Alpha의 value 프로퍼티 타입을 Float로 변경
그로 인한 다른 메서드들의 매개변수 타입 수정
Plane 구조체의 stepUpAlphaValue(), stepDownAlphaValue() 메서드에서 사용하는 newAlpha 상수의 값을 결정할때 Float 타입 값들끼리 뺄셈을 함으로 인해 소수의 값에 필요없는 값이 추가되어 해당 값 처리하는 과정을 추가

Hansolkkim committed:
Style: 오타 수정

Hansolkkim committed:
Refactor: 빈 칸 터치한 경우의 출력 흐름 수정

ViewController 클래스에 selectedView 프로퍼티 추가
Plane이 Rectangle을 특정하지 못 했을 때 Delegate에 알려주기 위해 nil을 사용할 수 있도록 PlaneDelegate의 rectangleDidSpecified(_:)의 매개변수를 Optional 타입으로 수정
rectangleDidSpecified(_:) 메서드에서 매개변수가 nil일 경우 initializeViewsInTouchedEmptySpaceCondition() 메서드를 호출하여 StatusView의 View들을 초기화하고 선택됐었던 RectangleView의 borderWidth가 0이 되도록 설정

Hansolkkim commented:
리뷰해주셨던 부분 모두 수정 완료했습니다! 
리뷰 내용대로 수정하던 도중 Delegate 메서드의 이름 규칙에 대해 궁금한 점이 생겼습니다.

스타일 가이드에는 Delegate의 메서드의 맨 앞에 오는 이벤트 주체의 이름과 제일 앞 매개변수가 일치해야 된다는 말이 있어서,
이번 스텝에서 제가 사용했던 메서드들은 대부분 Rectangle 객체에 대한 이벤트와 관련되있다고 생각했기 때문에 `rectangleDidAdded(_ rectangle: Rectangle)` 와 같이 사용했습니다.

그런데 다른 클래스에서 기본적으로 구현되어있는 Delegate 메서드에서는, 예를 들어 UITextFieldDelegate의 메서드는
`textFieldDidEndEditing(_ textField: UITextField..)` 의 형태로 UITextField 타입을 매개변수로 받도록 구현되어 있었습니다.

그래서 저는 이번 스텝에서의 제가 사용한 Delegate 메서드도 실질적으로는 Plane과 관련되어 있기 때문에 `plane-` 으로 시작되야 맞는 형태가 아닌가 생각이 들었습니다. 그래서 `planeDidAddRectangle(_)` 의 형태로 사용하려고 하니, 매개변수를 Plane 객체 타입으로 사용하는게 어색하게 느껴졌습니다.

혹시 이런 경우에는 컨벤션에 따르면 `rectangleDidAdded(_ rectangle: Rectangle)` , `planeDidAddRectangle(_ plane: Plane)` , 둘 중 어느 형태의 메서드가 더 적합한 형태인지가 적절한지 알 수 있을까요?

Hansolkkim committed:
Refactor: Plane 구조체의 changeBackgorundColor() 메서드 매개변수 추가

changeBackgroundColor() 메서드 이름과 자동으로 랜덤 색을 생성해 바꿔주는 동작이 맞지않다고 판단하여 ViewController에서 색상을 지정해서 넘겨주도록 수정

Hansolkkim committed:
Remove: Factory중 일부 제거

Factory 중 타입을 생성하는 팩토리들을 제거하고 각 타입에 random() 타입 메서드를 추가

godrm commented on DrawingApp/DrawingApp/Model/Alpha.swift:
```diff
@@ -8,27 +8,30 @@
-        } else if opacityLevel < min {
-            return nil
-        } else {
-            self.value = Double(opacityLevel) / Double(10)
-        }
-
+    init(opacityLevel: OpacityLevel) {
+        self.opacityLevel = opacityLevel
+        self.value = opacityLevel.rawValue
```
오 OpacityLevel을 내부에 enum으로 선언했군요. 
이런 경우 value를 OpacityLevel에서 계산하는 computed property로 해도 될 것 같네요

Hansolkkim replied:
처음에 그렇게 사용했다가, RectangleView에 alpha를 넣어주는 과정에서 `Alpha.opacityLevel.value ` 이렇게 호출하는게 복잡하다고 생각되어 위처럼 value를 다시 사용하는 방법으로 고쳤습니다. computed property로도 호출할때 이름이 어색하지 않는 방법이 없는지 고민해보겠습니다.

godrm commented on DrawingApp/DrawingApp/Model/Plane.swift:
```diff
@@ -19,45 +19,67 @@ struct Plane {
-                updateViewMatchedRectangleDelegate?.rectangleAlphaDidChanged(rectangle)
-                return rectangle
-            }
+    public func stepUpAlphaValue() -> Rectangle? {
```
changeAlphaValue()와 stepUpAlphaValue()는 뭐가 다른걸까요?
다 return 타입이 Rectangle? 인것도 좀 어색합니다. 뭔가 result가 필요한 게 아니라 전부 Rectangle이 필요한게 맞나요?

Hansolkkim replied:
> changeAlphaValue()와 stepUpAlphaValue()는 뭐가 다른걸까요?

다 같은 changeAlphaValue(to:) 메서드로 처리할 수 있겠네요 stepUpAlphaValue(), stepDownAlphaValue() 메서드 삭제하고 changeAlphaValue(to:) 메서드로 모두 처리하도록 하겠습니다.

> 다 return 타입이 Rectangle? 인것도 좀 어색합니다. 뭔가 result가 필요한 게 아니라 전부 Rectangle이 필요한게 맞나요?

테스트할 때 Rectangle? 을 반환받는 것이 적절하고, 반환값이 있어도 함수 동작에는 문제가 되지 않아 속성 변경 메서드들이 Rectangle? 타입을 반환하도록 했습니다.. 테스트 함수 짜는 부분이 아직 어려워 해당 부분을 처리할 방법을 더 고민해보고 수정해보겠습니다

godrm replied:
당연히 return 이 필요없거나 동작에 대한 결과를 주면 되는데, Rectangle?을 반환한다고 동작은 영향을 주지 않지만 이상한 겁니다
적절한 리턴값이 무엇인가를 고민해보세요. Result<> 라는 타입도 있습니다

Hansolkkim replied:
Result를 사용하면 Test에서도 더 잘 쓸 수 있겠네요,, 리턴값에 대해 고민해보고 수정해보겠습니다! 감사합니다!

godrm commented on DrawingApp/DrawingApp/Model/PlaneDelegate.swift:
```diff
@@ -0,0 +1,15 @@
+
+import Foundation
+
+protocol PlaneDelegate {
+    func rectangleDidAdded(_ rectangle: Rectangle)
+    func rectangleDidSpecified(_ rectangle: Rectangle?)
+    func rectangleBackgroundColorDidChanged(_ rectangle: Rectangle)
+    func rectangleAlphaDidChanged(_ rectangle: Rectangle)
```
프로토콜로 델리게이트를 정의할 때 PlaneDelegate 인데 이름 시작, 매개변수가 다 Rectangle인게 어색하네요.
이러면 RectangleDelegate여야 하지 않을까요?

Hansolkkim replied:
> 스타일 가이드에는 Delegate의 메서드의 맨 앞에 오는 이벤트 주체의 이름과 제일 앞 매개변수가 일치해야 된다는 말이 있어서,
이번 스텝에서 제가 사용했던 메서드들은 대부분 Rectangle 객체에 대한 이벤트와 관련되있다고 생각했기 때문에 rectangleDidAdded(_ rectangle: Rectangle) 와 같이 사용했습니다.
> 
> 그런데 다른 클래스에서 기본적으로 구현되어있는 Delegate 메서드에서는, 예를 들어 UITextFieldDelegate의 메서드는
textFieldDidEndEditing(_ textField: UITextField..) 의 형태로 UITextField 타입을 매개변수로 받도록 구현되어 있었습니다.
> 
> 그래서 저는 이번 스텝에서의 제가 사용한 Delegate 메서드도 실질적으로는 Plane과 관련되어 있기 때문에 plane- 으로 시작되야 맞는 형태가 아닌가 생각이 들었습니다. 그래서 planeDidAddRectangle(_) 의 형태로 사용하려고 하니, 매개변수를 Plane 객체 타입으로 사용하는게 어색하게 느껴졌습니다.
> 혹시 이런 경우에는 컨벤션에 따르면 rectangleDidAdded(_ rectangle: Rectangle) , planeDidAddRectangle(_ plane: Plane) , 둘 중 어느 형태의 메서드가 더 적합한 형태인지가 적절한지 알 수 있을까요?

@godrm 수정 후 댓글에서 질문드린 내용입니다. 델리게이트 메서드들의 이름을 설정하는 부분에 대해서 고민을 했는데 이번 리뷰 내용을 보니 더 궁금한 점이 많아졌습니다..ㅠㅠ
혹시 PlaneDelegate의 프로토콜명을 RectangleDelegate로 수정한다면, 해당 Delegate를 Plane에서 가지고 있어도 이상하지는 않을까요?

godrm replied:
맞아요. Plane에서 보내는 델리게이트면 PlaneDelegate가 맞겠죠. 
PlaneDelegate라면 Plane을 전달해줘야 하는게 아닐까요? 
RectangleDelegate이라면 Plane에서 호출하는 게 아니어야겠죠. 그러니까 이상하다는거죠 ㅎㅎ 
누가 보내느냐. 그렇지만 어떤 정보를 갖고 있느냐 그걸 혼동할 필요는 없습니다.

Hansolkkim replied:
답변 감사합니다!
그럼 Delegate 메서드들 이름을 모두 `plane-` 으로 시작하도록 수정하고, 매개변수도 (_ plane: Plane) 으로 수정하여 다음 PR때 추가하겠습니다!

godrm reviewed (approved):
개선된 방향에서 다음 스텝을 진행해보세요
